client/shared/src/search/query/parser.ts

```diff
@@ -123,21 +152,76 @@ export const parseParenthesis = (tokens: Token[]): State => {
     return { result: groupNodes.result, tokens }
 }
 
-export const parseLeaves = (tokens: Token[]): State => {
+const parseNot = (tokens: Token[]): State => {
+    const keyword = tokens[0]
+    tokens = tokens.slice(1) // consume NOT
+
+    let nodes: Node[] = []
+    const operand = tokens[0]
+
+    if (!operand) {
+        return { result: createOperator(nodes, OperatorKind.Not), tokens }
+    }
+
+    switch (operand.type) {
+        case 'openingParen':
+            const state = parseParenthesis(tokens)
+            if (state.result.type === 'error') {
+                return { result: state.result, tokens }
+            }
+            nodes = state.result.nodes
+            tokens = state.tokens
+            break
+        default:
+            const node = tokenToLeafNode(operand)
+            if (node.type === 'error') {
+                return { result: node, tokens }
+            }
+            nodes = node.nodes
+            tokens = tokens.slice(1)
+    }
+
+    return { result: createOperator(nodes, OperatorKind.Not, keyword.range.start), tokens }
+}
+
+/**
+ * parseSequence parses consecutive tokens. If the sequence has a size smaller
+ * than 2 the nodes are returned directly. That means there will never be a
+ * sequence of size 1.
+ * This also takes care of parsing the NOT keyword
+ */
+const parseSequence = (tokens: Token[]): State => {
     const nodes: Node[] = []
     while (true) {
         const current = tokens[0]
         if (current === undefined) {
             break
         }
         if (current.type === 'openingParen') {
-            return parseParenthesis(tokens)
+            const state = parseParenthesis(tokens)
+            if (state.result.type === 'error') {
+                return { result: state.result, tokens: state.tokens }
+            }
+            nodes.push(...state.result.nodes)
+            tokens = state.tokens
+            continue
         }
         if (current.type === 'closingParen') {
             break
         }
-        if (current.type === 'keyword' && (current.kind === KeywordKind.And || current.kind === KeywordKind.Or)) {
-            return { result: createNodes(nodes), tokens } // Caller advances.
+        if (current.type === 'keyword') {
+            if (current.kind === KeywordKind.And || current.kind === KeywordKind.Or) {
+                return { result: createSequence(nodes), tokens } // Caller advances.
```

Comment:
We use `break` for the `closingParen` case. I think behavior is more clear if we match that pattern here.

```suggestion
                break
```

Reply:
Makes sense... this line of code was there before, I don't know why it wasn't using `break` originally. Tests still pass.